ichnaea/async/tests.py

```diff
@@ -10,38 +9,85 @@
 
 class TestBeat(object):
 
-    def test_tasks(self, celery):
-        tmpdir = tempfile.mkdtemp()
-        filename = os.path.join(tmpdir, 'celerybeat-schedule')
+    def test_tasks(self, celery, tmpdir):
+        filename = str(tmpdir / 'celerybeat-schedule')
         beat_app = celery.Beat()
-        try:
-            beat = beat_app.Service(
-                app=celery, schedule_filename=filename)
-            signals.beat_init.send(sender=beat)
-            # parses the schedule as a side-effect
-            scheduler = beat.get_scheduler()
-            registered_tasks = set(scheduler._store['entries'].keys())
-        finally:
-            shutil.rmtree(tmpdir)
-
-        # Import tasks after beat startup, to ensure beat_init
-        # correctly loads configured Celery imports.
+        beat = beat_app.Service(app=celery, schedule_filename=filename)
+        signals.beat_init.send(sender=beat)
+
+        # Parses the schedule as a side-effect
+        scheduler = beat.get_scheduler()
+        registered_tasks = set(scheduler._store['entries'].keys())
+
+        # Import tasks after beat startup, to ensure beat_init correctly loads
+        # configured Celery imports.
         from ichnaea.data import tasks
-        all_tasks = set([m[1].shortname() for m in getmembers(tasks)
-                         if isinstance(m[1], BaseTask)])
-
-        assert (all_tasks - registered_tasks ==
-                set(['data.update_blue', 'data.update_cell',
-                     'data.update_wifi',
-                     'data.update_datamap', 'data.cleanup_datamap',
-                     'data.cell_export_diff', 'data.cell_export_full',
-                     'data.export_reports', 'data.update_statregion',
-                     ]))
-
-        assert (set(['_'.join(name.split('_')[:-1]) for name in
-                     registered_tasks - all_tasks]) ==
-                set(['data.update_blue', 'data.update_cell',
-                     'data.update_wifi']))
+        all_tasks = set([
+            m[1].shortname() for m in getmembers(tasks)
+            if isinstance(m[1], BaseTask)
+        ])
+
+        assert (
+            all_tasks - registered_tasks
+            == set([
+                'data.update_blue',
+                'data.update_cell',
+                'data.update_wifi',
+                'data.update_datamap',
+                'data.cleanup_datamap',
+                'data.cell_export_diff',
+                'data.cell_export_full',
+                'data.export_reports',
```

Comment:
For some reason, this PR changes the output of this specific check. Before `data.update_statregion` would be in `all_tasks`, but not `registered_tasks`. Now it's in both sets.

I spent some time trying to figure out what changed and why this test broke, but I'm completely puzzled as to why it shouldn't show up in `registered_tasks` before. My current theory is that it's a bug in the test and `data.update_statregion` should have been in `registered_tasks` and something got fixed when I did this PR.

Reply:
It looks like on master these tasks are not registered, which are now registered with this PR:

```
data.cleanup_datamap_ne
data.cleanup_datamap_nw
data.cleanup_datamap_se
data.cleanup_datamap_sw
data.update_datamap_ne
data.update_datamap_nw
data.update_datamap_se
data.update_datamap_sw
data.update_statregion
```

These all have decorators with the attribute ``_enabled=_web_content_enabled``, so in master that is ``False`` and in this branch it is ``True`` when the task decorators are evaluated. ``data.update_statregion`` is special because it doesn't have a ``_shard_model``, so the name is the same after registration.

This test is checking:
* Does the `_enabled` parameter to ``@celery_app.task`` work, skipping registration when ``False``? It looks like one value flipped from ``False`` to ``True``, breaking the test in a non-obvious way.
* Does the `shard_model` parameter to `@celery_app.task` work, creating many tasks that vary by a suffix?
* Did we get the expected all and registered tasks, but not by looking directly, but by comparing differences of sets?

This test seems problematic to me. Maybe it just makes sense to test ``all_tasks`` and ``registered_tasks`` directly, rather than try to test the set differences and loop through looking for shard suffixes. Or, it should setup the environment to ensure what tests are enabled, or should vary the expected value by ``_web_content_enabled`` and ``_cell_export_enabled``. Or maybe this shouldn't be a unit test.

Reply:
Aha! That's fantastic! Thank you for figuring that out.

I think I'm going to tackle #871 which changes the whole "web content enabled/disabled" thing. I'll leave this for now and address it then.